File: examples/Clojure/taskvent.clj
;; Task ventilator
;; Binds PUSH socket to tcp://localhost:5557
;; Sends batch of tasks to workers via that socket

(ns taskvent
  (:require [zeromq.zmq :as zmq]))

(defn work [socket]
  ;; Random workload from 1 to 100 ms
  (let [workload (inc (rand-int 100))]
    (zmq/send-str socket (str workload))
    workload))

(defn -main []
  (let [context (zmq/zcontext 1)]
    (with-open [sender (doto (zmq/socket context :push)
                         (zmq/bind "tcp://*:5557"))
                sink (doto (zmq/socket context :push)
                       (zmq/connect "tcp://127.0.0.1:5558"))]
    (println "Press Enter when the workers are ready: ")
    (read-line)
    (println "Sending tasks to workers…")
      (zmq/send-str sink "0")
      (let [times (repeatedly 100 (partial work sender))]
        (printf "Total expected cost: %d msec\n" (apply + times))))))

File: examples/Go/taskvent.go
//
// Task ventilator
// Binds PUSH socket to tcp://localhost:5557
// Sends batch of tasks to workers via that socket
//
package main

import (
	"fmt"
	zmq "github.com/alecthomas/gozmq"
	"math/rand"
	"time"
)

func main() {
	context, _ := zmq.NewContext()
	defer context.Close()

	// Socket to send messages On
	sender, _ := context.NewSocket(zmq.PUSH)
	defer sender.Close()
	sender.Bind("tcp://*:5557")

	//  Socket to send start of batch message on
	sink, _ := context.NewSocket(zmq.PUSH)
	defer sink.Close()
	sink.Connect("tcp://localhost:5558")

	fmt.Print("Press Enter when the workers are ready: ")

	var line string
	fmt.Scanln(&line)

	fmt.Println("Sending tasks to workers…")

	sink.Send([]byte("0"), 0)

	// Seed the random number generator
	rand.Seed(time.Now().UnixNano())

	total_msec := 0

	for i := 0; i < 100; i++ {
		workload := rand.Intn(100)
		total_msec += workload
		msg := fmt.Sprintf("%d", workload)
		sender.Send([]byte(msg), 0)
	}

	fmt.Printf("Total expected cost: %d msec\n", total_msec)

	time.Sleep(1e9) //  Give 0MQ time to deliver: one second ==  1e9 ns

}
